### PR TITLE
Add Capability for Coupling FE Variables to FV Kernels.

### PR DESCRIPTION
**Pull request by csdechant**
## Reason
This PR is to address the issue for FE to FV coupling in #15062 and fixing git history in #18352.

## Impact
This PR will allow user to couple the element / face average values for FE variables and gradients to FV element / flux kernels.

### CODE REVIEW

File: framework/include/variables/MooseVariableFE.h [hunk 1]
```diff
@@ -52,6 +52,10 @@ InputParameters validParams<ArrayMooseVariable>();
 template <typename OutputType>
 class MooseVariableFE : public MooseVariableField<OutputType>
 {
+private:
+  ADTemplateVariableGradient<OutputType> _ad_grad_u_face;
+  ADTemplateVariableGradient<OutputType> _ad_grad_neighbor_u_face;
```
Comment (lindsayad) [inline framework/include/variables/MooseVariableFE.h:57]: We generally try to order class definitions like this:
```c++
public:
protected:
private:
```

File: framework/include/variables/MooseVariableFE.h [hunk 2]
```diff
@@ -473,6 +477,13 @@ class MooseVariableFE : public MooseVariableField<OutputType>
   virtual void computeElemValues() override;
   virtual void computeElemValuesFace() override;
   virtual void computeFaceValues(const FaceInfo & fi) override;
+
+  /// Adding a adGradSln that is the average of the face gradients on both sides
+
+  VectorValue<ADReal> _ad_grad_face_avg;
+  void computeAdGradFaceAvg(const FaceInfo & fi);
+  VectorValue<ADReal> adGradSln(const FaceInfo & fi);
```
Comment (lindsayad) [inline framework/include/variables/MooseVariableFE.h:485]: We also generally try to put methods before data. Additionally data should tend to be less accessible than methods. For instance I suspect that `_ad_grad_face_avg` could probably be made private

File: framework/src/fvkernels/FVFluxKernel.C [hunk 3]
```diff
@@ -300,3 +300,18 @@ FVFluxKernel::gradUDotNormal() const
...
+  const auto * _coupled_check = getVarHelper<MooseVariableField<Real>>(name, 0);
+
+  if (_coupled_check->isFV())
+  {
+    const MooseVariableFV<Real> * _coupled_var = dynamic_cast<const MooseVariableFV<Real> *>(getFieldVar(name, 0));
+    return _coupled_var->adGradSln(*_face_info);
+  }
+
+  MooseVariable & _coupled_var = *getVar(name, 0);
+  return _coupled_var.adGradSln(*_face_info);
```
Comment (lindsayad) [inline framework/src/fvkernels/FVFluxKernel.C:316]: If it's not there already, we should add a pure virtual `adGradSln(const FaceInfo &)` method in `MooseVariableField`. Then this method's code could be simplified all the way down to
```c++
_coupled_check->adGradSln(*_face_info);
```
And then with that simplification I would probably just rename `_coupled_check` to `_coupled_var`

**Review by lindsayad — CHANGES_REQUESTED** [review]
Just took a glance at this one commit. I see now how you implemented the coupled face gradient method; it's on-the-fly which is what I thought it would need to be.

I believe we will also want to be able to call something similar for FVFluxBCs. It may be worth moving `adCoupledGradientFace` to `Coupleable` and then add a `FaceInfo` argument so you can pass in `_face_info` (since `Coupleable` doesn't have access to that data member)

**Comment by csdechant** [comment]
@lindsayad I think I addressed all your recommended changes and I added some test for the FE to FV coupling. Please let me know if you think anything else needs to be changed or added.

**Comment by lindsayad** [comment]
@csdechant what is the largest of your tests? I want to profile one of the coupling tests

**Comment by lindsayad** [comment]
Per discussion with @friedmud I think we want to modify the coupled gradient method to take an `unsigned int` corresponding to the variable number instead of taking the string name. Then we don't have to hash strings for map lookup during residual/Jacobian evaluation. We should determine the `unsigned int` during object construction using the `coupled` method (in `Coupleable`).

Other ideas would be to 
1. bind `&MooseVariableField<Real>::adGradSln` to a function pointer (or `std::function`) data member in the kernel object at construction...but then you would need an instance of the variable to call it. This leads me to...
2. Remove the `Coupleable::adCoupledGradientFace(const std::string & var_name, const FaceInfo & fi)` API altogether which would force kernel developers to get the variable and call the variable method directly. This would incur no map or vector lookups

**Comment by lindsayad** [comment]
@csdechant are you going to continue to work on this PR or would you like me to take it over? I'll definitely leave it to you if you want it!

I'm thinking that maybe we should take the new Coupleable API out for now and we can always add it back if users want the convenience (at the sacrifice of speed). I would do this in a new commit so that we can easily restore the API with a `git revert` in the future if we want

**Comment by csdechant** [comment]
@lindsayad Sorry for the late response, (I just notice I wasn't receiving notifications for this PR, but that should be fixed). To answer your questions: 
- The largest test was `2D_Coupling_FEGrad_to_FVFluxKernel.i`, but all the test had similar run times.
- I can take the Coupleable API out as a new commit. I have been busy with getting ready for GEC and finish a paper, but I will more time to focus on this PR after GEC.

**Comment by lindsayad** [comment]
> I can take the Coupleable API out as a new commit. I have been busy with getting ready for GEC and finish a paper, but I will more time to focus on this PR after GEC.

Sounds good. By the time you return to it we may leverage some of the ideas in #18395 and #18806

**Comment by cticenhour** [comment]
@csdechant since I think this is still being worked on, I am posting a message to stop the stalebot from closing this.

**Pull request closed without merging**